src/v/storage/segment_utils.cc

```diff
@@ -510,9 +510,9 @@ ss::future<std::optional<size_t>> do_self_compact_segment(
   storage::probe& pb,
```

Comment:
```
T: Take write lock.
T: Truncate the segment, deleting the compacted index.
C: Take read lock.
C: Attempt to self compact, but compacted index doesn't exist.

This commit fixes this by taking the read lock before the first step,
```

Did you consider having truncate take the write lock, instead of holding the read lock for a larger span? I don't recall all the locking details, but I do remember it wasn't very intuitive with naming.

Reply:
Sorry this isn't clear, the write lock is held during the truncation (imagine each lock holds for the next line in this sequence)